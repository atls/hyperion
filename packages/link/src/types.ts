export interface NextLinkProps {
  fontFamily?: string
  fontWeight?: string | number | string[] | number[]
  lineHeight?: string | number | string[] | number[]
  fontSize?: string | number | string[] | number[]
  color?: string
  hoverColor?: string
  clickedColor?: string
  href: string
  target?: string
  display?: string
  as?: string
  query?: any
  textTransform?: string
  textAlign?: string
  underline?: boolean
  width?: string
  whiteSpace?: string
  dangerouslySetInnerHTML?: any
  textDecorationStyle?: string
  textDecorationColor?: string
  passHref?: boolean
  children?: any
}

export interface LinkProps {
  children?: any
  href?: string
  fontWeight?: string | number | string[] | number[]
  lineHeight?: string | number | string[] | number[]
  fontSize?: string | number | string[] | number[]
  fontFamily?: string
  whiteSpace?: string
  dangerouslySetInnerHTML?: any
  color?: string
  hoverColor?: string
  textDecorationStyle?: string
  textDecorationColor?: string
  clickedColor?: string
  textTransform?: string
  underline?: boolean
  target?: string
  width?: string
  pt?: string
  onClick?: any
  onMouseOver?: () => void
  onMouseLeave?: () => void
}
