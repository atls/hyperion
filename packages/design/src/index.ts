export * from './Sample'
export * from './Layout'
export * from './Container'
export * from './List'
export * from './Text'
export * from './Space'
