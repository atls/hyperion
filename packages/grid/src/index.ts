export * from './Column'
export * from './Grid'
