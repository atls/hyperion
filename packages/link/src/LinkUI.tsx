import { layout }   from 'styled-system'
import { ifProp }   from 'styled-tools'

import styled       from '@emotion/styled'
import { Text }     from '@atlantis-lab/text'
import { useTheme } from '@emotion/react'

export const LinkUI = styled<any>(Text.withComponent('a'))(
  {
    position: 'relative',
    textDecoration: 'none',
    display: 'inline-flex',
    transition: '0.35s',
    cursor: 'pointer',
    '::after': {
      transition: '0.35s',
      content: '""',
      height: '1px',
      width: '100%',
      backgroundColor: 'transparent',
      position: 'absolute',
      bottom: '-3px',
      left: '0px',
    },
  },
  ifProp('underline', ({ underlineColor }) => {
    const theme: any = useTheme()

    return {
      position: 'relative',
      '::after': {
        transition: '0.35s',
        content: '""',
        height: '1px',
        width: '100%',
        backgroundColor: theme.colors[underlineColor] || theme.colors.lightBlue,
        position: 'absolute',
        bottom: '-3px',
        left: '0px',
      },
    }
  }),
  layout,
)
