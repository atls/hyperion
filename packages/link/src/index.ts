export * from './Link'
export * from './NextLink'
export * from './RouteLink'
