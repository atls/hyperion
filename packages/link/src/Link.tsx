/* eslint-disable react/require-default-props */

import React, { FC, useState } from 'react'

import { withTheme }           from '@emotion/react'

import { LinkUI }              from './LinkUI'

interface LinkProps {
  children?: any
  href?: string
  fontWeight?: string | number | string[] | number[]
  lineHeight?: string | number | string[] | number[]
  fontSize?: string | number | string[] | number[]
  fontFamily?: string
  whiteSpace?: string
  dangerouslySetInnerHTML?: any
  color?: string
  hoverColor?: string
  underlineColor?: string
  clickedColor?: string
  textTransform?: string
  underline?: boolean
  target?: string
  width?: string
  pt?: string
  onClick?: any
  onMouseOver?: () => void
  onMouseLeave?: () => void
}

export const Link: FC<LinkProps> = React.forwardRef(
  (
    {
      children,
      href,
      fontSize,
      lineHeight,
      fontWeight,
      color,
      hoverColor,
      clickedColor,
      underline,
      underlineColor,
      ...props
    },
    ref,
  ) => {
    const [hovered, setHovered] = useState(false)
    const [clicked, setClicked] = useState(false)
    return (
      <LinkUI
        fontSize={fontSize}
        lineHeight={lineHeight}
        ref={ref}
        fontWeight={fontWeight}
        /* eslint-disable no-nested-ternary */
        color={clicked ? clickedColor || color : hovered ? hoverColor || color : color}
        href={href}
        underline={underline && hovered && ((!clicked && clickedColor) || !clickedColor)}
        underlineColor={underlineColor}
        onMouseOver={() => setHovered(true)}
        onFocus={() => setHovered(true)}
        onMouseLeave={() => {
          setHovered(false)
          setClicked(false)
        }}
        onMouseDown={() => setClicked(true)}
        onMouseUp={() => setClicked(false)}
        {...props}
      >
        {children}
      </LinkUI>
    )
  },
)
