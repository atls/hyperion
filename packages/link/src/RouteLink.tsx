import React        from 'react'
import { Link }     from 'react-router-dom'

import { useTheme } from '@emotion/react'

import { LinkUI }   from './LinkUI'

const RouteLinkStyled = LinkUI.withComponent(Link)

export const RouteLink = (props) => {
  const theme = useTheme()

  return <RouteLinkStyled theme={theme} {...props} />
}
