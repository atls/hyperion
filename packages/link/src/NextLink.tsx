/* eslint-disable react/require-default-props */

import Link          from 'next/link'
import React         from 'react'

import { withTheme } from '@emotion/react'

import { LinkUI }    from './LinkUI'

interface NextLinkProps {
  fontFamily?: string
  fontWeight?: string | number | string[] | number[]
  lineHeight?: string | number | string[] | number[]
  fontSize?: string | number | string[] | number[]
  color?: string
  hoverColor?: string
  clickedColor?: string
  href: string
  target?: string
  display?: string
  as?: string
  query?: any
  textTransform?: string
  textAlign?: string
  width?: string
  whiteSpace?: string
  dangerouslySetInnerHTML?: any
  underline?: boolean
  passHref?: boolean
  children?: any
}

export const NextLink = ({
  href,
  children,
  as,
  query,
  passHref = true,
  ...props
}: NextLinkProps) => (
  <Link href={{ pathname: href, query: query || {} }} as={as || href} passHref={passHref}>
    <LinkUI {...props}>{children}</LinkUI>
  </Link>
)

NextLink.defaultProps = {
  fontFamily: 'primary',
  fontWeight: '',
}
