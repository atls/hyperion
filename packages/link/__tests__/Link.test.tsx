import React             from 'react'
import renderer          from 'react-test-renderer'

import { ThemeProvider } from '@emotion/react'

import { Link }          from '../src'

describe('test suit for link', () => {
  it('should render Link correctly', () => {
    const defaultTheme = {
      colors: {
        white: '#FFF',
        lightBlue: '#00F',
      },
    }

    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Link href='/nowhere'>Click</Link>
        </ThemeProvider>,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
