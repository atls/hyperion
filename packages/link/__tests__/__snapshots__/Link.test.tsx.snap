// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test suit for link should render Link correctly 1`] = `
.emotion-0 {
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: 0.35s;
  transition: 0.35s;
  cursor: pointer;
}

.emotion-0::after {
  -webkit-transition: 0.35s;
  transition: 0.35s;
  content: "";
  height: 1px;
  width: 100%;
  background-color: transparent;
  position: absolute;
  bottom: -3px;
  left: 0px;
}

<a
  className="emotion-0"
  href="/nowhere"
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
>
  Click
</a>
`;
